rust/connlib/tunnel/src/lib.rs

```diff
@@ -110,53 +78,282 @@ pub(crate) fn get_v6(ip: IpAddr) -> Option<Ipv6Addr> {
+                            .untransform(packet.source(), self.write_buf.as_mut())
+                            .ok();
+                    }
+                    Ok(None) => None,
+                    Err(e) => {
+                        tracing::error!(%local, %from, "Failed to decapsulate incoming packet: {e}");
+                        None
+                    }
+                }
+            }
+            (_, Err(e)) => {
+                tracing::error!("Failed to read socket: {e:#?}");
+                None
+            }
+        }
+    }
+}
+
+struct IceSockets {
+    udp_sockets: UdpSockets<MAX_UDP_SIZE>,
+    relay_socket_v4: Option<Socket<MAX_UDP_SIZE>>,
+    relay_socket_v6: Option<Socket<MAX_UDP_SIZE>>,
+}
+
+impl IceSockets {
+    fn socket_send(&mut self, transmit: &Transmit) -> Result<usize> {
+        match (transmit.src, transmit.dst) {
+            (Some(src), _) => {
+                Ok(self
+                    .udp_sockets
+                    .try_send_to(src, transmit.dst, &transmit.payload)?)
+            }
+            (None, SocketAddr::V4(_)) => {
+                let socket = self.relay_socket_v4.as_ref().ok_or(Error::NoIpv4)?;
+                Ok(socket.try_send_to(transmit.dst, &transmit.payload)?)
+            }
+            (None, SocketAddr::V6(_)) => {
+                let socket = self.relay_socket_v6.as_ref().ok_or(Error::NoIpv6)?;
+                Ok(socket.try_send_to(transmit.dst, &transmit.payload)?)
+            }
+        }
+    }
+
+    fn poll_recv_from<'a>(
+        &'a mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<(SocketAddr, io::Result<(SocketAddr, &'a mut [u8])>)> {
+        if let Poll::Ready(packet) = self.udp_sockets.poll_recv_from(cx) {
+            return Poll::Ready(packet);
+        }
+
+        if let Some(Poll::Ready(packet)) =
+            self.relay_socket_v4.as_mut().map(|s| s.poll_recv_from(cx))
+        {
+            return Poll::Ready(packet);
+        }
+
+        if let Some(Poll::Ready(packet)) =
+            self.relay_socket_v6.as_mut().map(|s| s.poll_recv_from(cx))
+        {
+            return Poll::Ready(packet);
+        }
+
+        Poll::Pending
+    }
+}
+
+struct ConnectionState<TRole, TId, TTransform> {
+    connections: Connections<TRole, TId, TTransform>,
+    connection_pool_timeout: BoxFuture<'static, std::time::Instant>,
+    if_watcher: IfWatcher,
+    sockets: IceSockets,
+}
+
+impl<TRole, TId, TTransform> ConnectionState<TRole, TId, TTransform>
+where
+    TId: Eq + Hash + Copy + fmt::Display,
+    TTransform: PacketTransform,
+{
+    fn new(private_key: StaticSecret) -> Self {
+        let if_watcher = IfWatcher::new().expect(
+            "Program should be able to list interfaces on the system. Check binary's permissions",
+        );
+
+        let mut connection_pool = Node::new(private_key, std::time::Instant::now());
+        let mut udp_sockets = UdpSockets::default();
+
+        for ip in if_watcher.iter() {
```

Comment:
Can we do something better than this? This would give us 1 local candidate per ip. This can be a problem due to the combinatoric number of candidate pairs, that's why we used to limit it to 8. I'd like to at least limit this to 1 ipv4 and 1 ipv6 per iface but we don't have which iface the ip came from here.

Reply:
I'll do some research on this today.